### ISSUE TITLE
Option to accept invalid Content-Encoding header / Disable compression

### ISSUE THREAD

**Issue by daboross** [actix/actix-web#350]
I'm rewriting a python/flask web server into rust/actix, and running into one issue: some of the clients send invalid `Content-Encoding` headers.

Specifically, an application I wrote incorrectly sends `Content-Encoding: gzip` without actually compressing the data being sent. The python/flask server simply ignored the invalid content-encoding header, so this bug was left undiscovered.

Actix correctly refuses it, and outputs:

```
[WARN][actix_web::pipeline] Error occured during request handling: Error that occur during reading payload: invalid gzip header
```

I'd like to just drop support for my old client, but unfortunately I can't. (this is a version usage statistics service, and would be fairly useless if it ignored old clients)

Right now I do have a solution, but it isn't an ideal one. If I set `default-features = false` for `actix-web` to disable the `brotly` and `flate` features, actix does not attempt to decompress the data, and it fully works.

Ideally there would be some way to accept invalid headers like this on a specific resource. However, I realize my use case is a bit niche, and I doubt anyone else has a client that makes this bad of a request.

**Comment by DoumanAsh** [comment]
While it is understandable but I'm not sure if it is good idea to pander bad clients in general?
The problem is also why server should assume data is uncompressed while client specified compression?

The approach you're suggesting is making `actix-web` very keen to handle this particular use case, but what if client would send data with different compression while specificing `gzip`? 
Considering possible content encodings there are multiple assumptions we can take, and neither can be correct.
Any conditional ignore of headers also imply performance degradation.

**Comment by DJMcNab** [comment]
Is it possible to use a middleware to fix this, or is the request translated before any middleware are run.

**Comment by DoumanAsh** [comment]
@DJMcNab yes, https://actix.rs/actix-web/actix_web/middleware/trait.Middleware.html#method.start should run when actix-web receives request. AFAIK it should not attempt to read response before user himself will try to get to read it

**Comment by kosta** [comment]
What about the following approach:
Install your own middleware (before the builtin ones) that looks at the two few bytes and check if they are 1F 8B if the Content-Encoding header is gzip. If they are not, remove the Content-Encoding header.

**Comment by daboross** [comment]
> While it is understandable but I'm not sure if it is good idea to pander bad clients in general?

Definitely agree that we shouldn't change the general behavior. If it weren't for backwards compatibility in my specific situation, I would 100% go with actix's current behavior.

@kosta 

> What about the following approach:
> Install your own middleware (before the builtin ones) that looks at the two few bytes and check if they are 1F 8B if the Content-Encoding header  is gzip. If they are not, remove the Content-Encoding header.

This sounds like the best solution- thanks! I'll try to research and find a way to peek at the first two bytes without consuming them.

**Comment by DoumanAsh** [comment]
@daboross If your client uses agent header then you could use it to identify how to handle it, otherwise aside from peeking into content I don't see any other way.

Feel free to close issue if you got answer to your question.
And for future you can use our gitter room to get any sort of consultation/help like that

**Comment by daboross** [comment]
@DoumanAsh Thanks! I will definitely come on gitter in the future- just opened this up since i wasn't sure if it could be solved without changes to actix-web.

Unfortunately there isn't any identifying thing that doesn't apply to later versions which send correct requests. I don't think any other clients use gzip encoding either, though, so removing the Content-Encoding header for all requests going through this particular middleware would be alright.

For actually doing that: do you know the best way to modify headers from within a middleware? The only header-related method I can find on [`HttpRequest`](https://actix.rs/actix-web/actix_web/struct.HttpRequest.html) is [`HttpMessage::headers`](https://actix.rs/actix-web/actix_web/trait.HttpMessage.html#tymethod.headers), which is read-only.

**Comment by DoumanAsh** [comment]
Ah... I'm sorry, I thought there was such method but it seems actually we do not provide way for user to modify request.
This is not nice.
But I think during major refactoring #348 we can get it

**Comment by fafhrd91** [comment]
i think we should add application or server wide option to disable request decompression.

at the moment it is not possible to alter behavior in runtime. when App receives `Request` decompression config is already done.

**Comment by fafhrd91** [comment]
compression/decompression feature is implemented as middlewares in 1.0, it is possible to create custom middleware or provide PR
